context.go

```diff
@@ -37,16 +37,13 @@ func NewContext(app *App, set *flag.FlagSet, parentCtx *Context) *Context {
 
 	if c.Context == nil {
 		ctx, cancel := context.WithCancel(context.Background())
+		signals := make(chan os.Signal, 1)
 		go func() {
-			// wait for context cancellation
-			<-ctx.Done()
-			app.InterruptHandlerFunc(c)
-		}()
-		go func() {
-			defer cancel()
-			signals := make(chan os.Signal, 1)
-			signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
-			<-signals
+			for {
+				signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+				<-signals
+				c.App.InterruptHandler(c, cancel)
```

Comment:
I am still wonder if the API is cleaner if we add a ctx.Cancel() and just pass a single thing, what do you think?

Reply:
The API is clean either way. What we need to decide is how much of power to hand over to the user? If we call the cancel function ourselves, the user would not be able to do things like wait a certain number of times before processing the `SIGINT`. 

Consider this example.  The `docker-compose` CLI has two methods of handling the `SIGINT`. When we do a `Ctrl+C` the first time, the CLI would start shutting down the containers gracefully. However if we do a `Ctrl+C` the second time, it would immediately and forcibly shutdown the containers.

If we wish that our users build similar behaviour, I think we need to give them the cancel function and let them decide when to cancel the Context.

Reply:
Sure, but that's just custom interrupt handler which the user can implement.

What I am arguing for is instead of passing cancellation function as an argument, we expose the cancellation function as a function of the context, so the user can cancel ot from wherever he wants.

Reply:
There is no need to expose the function. The `cli.Context` embeds `context.Context`. So, anywhere the user is using the `cli.Context` they can simply call the `Done()` method to get access to the channel.

I think I understand where you are coming from. Let me work on this and let's discuss this again tomorrow

Reply:
Done is a method that returns a channel the user should check, it does not terminate the context.